packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts

```diff
@@ -109,248 +111,149 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
-          !(await this._imodelAccess.classDerivesFrom(identifier.className, className)) &&
-          !(await this._imodelAccess.classDerivesFrom(className, identifier.className))
-        ) {
-          continue;
-        }
-      }
-      if (!filteredChildrenIdentifierPaths) {
-        filteredChildrenIdentifierPaths = [];
-      }
-      if (path.length > 1) {
-        filteredChildrenIdentifierPaths.push(options ? { path: path.slice(1), options } : path.slice(1));
-      } else if (isFilterTarget) {
-        filterTargetOptions = HierarchyFilteringPath.mergeOptions(filterTargetOptions, options);
-      } else {
-        isFilterTarget = true;
-        filterTargetOptions = options;
-      }
-    }
-    return { filteredChildrenIdentifierPaths, filterTargetOptions, isFilterTarget };
+
+      return parsedNode;
+    };
   }
 
   public async defineHierarchyLevel(props: DefineHierarchyLevelProps): Promise<HierarchyLevelDefinition> {
     const sourceDefinitions = await this._source.defineHierarchyLevel(props);
-    const filteringProps = extractFilteringProps(this._nodeIdentifierPaths, props.parentNode);
-    if (!filteringProps) {
+
+    const filteringHelper = createHierarchyFilteringHelper(this._nodeIdentifierPaths, props.parentNode);
+    const childNodeFilteringIdentifiers = filteringHelper.getChildNodeFilteringIdentifiers();
+    if (!childNodeFilteringIdentifiers) {
       return sourceDefinitions;
     }
 
-    const filteredDefinitions: HierarchyLevelDefinition = [];
-    await Promise.all(
-      sourceDefinitions.map(async (definition) => {
-        let matchedDefinition: HierarchyNodesDefinition | undefined;
-        if (HierarchyNodesDefinition.isGenericNode(definition)) {
-          matchedDefinition = await matchFilters<GenericNodeKey>(
-            definition,
-            filteringProps,
-            async (id) => {
-              return (
-                HierarchyNodeIdentifier.isGenericNodeIdentifier(id) &&
-                (!id.source || id.source === this._imodelAccess.imodelKey) &&
-                id.id === definition.node.key
-              );
-            },
-            this._imodelAccess,
-            (def, matchingFilters) => {
-              const filteredChildrenIdentifierPaths = matchingFilters.reduce(
-                (r, c) => [...r, ...c.childrenIdentifierPaths],
-                new Array<HierarchyFilteringPath>(),
-              );
+    const [genericNodeDefinitions, instanceNodeDefinitions] = partition(sourceDefinitions, HierarchyNodesDefinition.isGenericNode);
+    return firstValueFrom(
+      merge(
+        genericNodeDefinitions.pipe(
+          map((definition) => {
+            if (
+              filteringHelper.hasFilterTargetAncestor ||
+              childNodeFilteringIdentifiers
+                .filter((id) => HierarchyNodeIdentifier.isGenericNodeIdentifier(id) && (!id.source || id.source === this._imodelAccess.imodelKey))
+                .some(({ id }) => id === definition.node.key)
```

Comment:
Why do we first filter, and then do `some`. I think we can immediatly do `some`. For example:
```ts
 if (
    filteringHelper.hasFilterTargetAncestor ||
    childNodeFilteringIdentifiers.some(
      (id) =>
        HierarchyNodeIdentifier.isGenericNodeIdentifier(id) &&
        (!id.source || id.source === this._imodelAccess.imodelKey) &&
        id.id === definition.node.key,
    )
  ) {
            ```

Reply:
Agree, changed.